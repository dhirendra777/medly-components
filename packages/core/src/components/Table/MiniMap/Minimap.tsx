import React, { FC, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { Minimap as StyledMinimap, MinimapWrapper, SliderContent, SliderContentWrapper } from './Minimap.styled';
import { SliderController } from './SliderController/SliderController';
import { Props } from './types';

export const Minimap: FC<Props> = React.memo(({ minimapWidth, controllerWidth, sliderContentPadding, tableRef }) => {
    const sliderControllerRef = useRef(null),
        wrapperRef = useRef(null),
        [mouseDown, setMouseDown] = useState(false),
        [oneMinimapToTableWidth, setOneMinimapToTableWidth] = useState(0) /* Table width to slider width ratio */,
        [oneTableToMinimapWidth, setOneTableToMinimapWidth] = useState(0) /* Slider width to table width ratio  */,
        [sliderWidth] = useState(minimapWidth - controllerWidth - sliderContentPadding * 2),
        [minimapOffsetTop, setMinimapOffsetTop] = useState({ left: 0, bottom: 0 }),
        [isHorizontalScrollPresent, setIsHorizontalScrollPresent] = useState(false),
        sliderContent = useMemo(
            () => (
                <SliderContentWrapper sliderWidth={minimapWidth}>
                    {new Array(5).fill(null).map((_, index) => (
                        <SliderContent key={index} controllerWidth={controllerWidth} />
                    ))}
                </SliderContentWrapper>
            ),
            [controllerWidth]
        );
    const isTableTallerThanWindow = useCallback(() => {
            const documentElementClientRect = document.documentElement.getBoundingClientRect();
            console.log('Taller than Data is ', tableRef.current.getBoundingClientRect().bottom, documentElementClientRect);
            return (
                tableRef.current &&
                tableRef.current.getBoundingClientRect().bottom > documentElementClientRect.y + documentElementClientRect.height
            );
        }, [tableRef.current, document.documentElement]),
        positionMinimap = useCallback(
            (isTableHeightGreaterThanWindow: boolean) => {
                const minimapLeftOffset = 170; /* Minimum left offset for minimap left positioning */
                const minimumBottomOffset = 95; /* Minimum bottom offset for minimap bottom positioning */
                let offsetToSet = minimumBottomOffset;
                if (isTableHeightGreaterThanWindow) {
                    offsetToSet =
                        tableRef.current.getBoundingClientRect().bottom -
                        document.documentElement.getBoundingClientRect().height +
                        minimumBottomOffset;
                    if (offsetToSet < minimumBottomOffset) offsetToSet = minimumBottomOffset;
                }
                return { left: tableRef.current && tableRef.current.clientWidth - minimapLeftOffset, bottom: offsetToSet };
            },
            [tableRef.current]
        ),
        setMinimapDimensions = useCallback(() => {
            const isTableTallerThanWindowResult = isTableTallerThanWindow();
            const tableW = (tableRef.current && tableRef.current.scrollWidth - tableRef.current.offsetWidth) + 1;
            setOneMinimapToTableWidth(tableW / sliderWidth);
            setOneTableToMinimapWidth(sliderWidth / tableW);
            setMinimapOffsetTop(positionMinimap(isTableTallerThanWindowResult));
            console.log('Data is ', tableRef.current && tableRef.current.scrollWidth > tableRef.current.offsetWidth);
            setIsHorizontalScrollPresent(tableRef.current && tableRef.current.scrollWidth > tableRef.current.offsetWidth);
        }, [sliderWidth, isTableTallerThanWindow, tableRef.current]),
        onTableScroll = useCallback(
            (e: React.UIEvent<HTMLDivElement>) => {
                e.stopPropagation();
                if (mouseDown) return false;
                sliderControllerRef.current.style.left = `${tableRef.current.scrollLeft * oneTableToMinimapWidth + sliderContentPadding}px`;
            },
            [sliderControllerRef.current, tableRef.current, oneTableToMinimapWidth, mouseDown]
        ),
        onWindowScroll = useCallback(
            (e: any) => {
                e.stopPropagation();
                setMinimapOffsetTop(positionMinimap(isTableTallerThanWindow()));
            },
            [positionMinimap, isTableTallerThanWindow]
        ),
        positionSliderController = useCallback(
            (e: any) => {
                e.stopPropagation();
                const clientX = e.touches ? e.touches[0].clientX : e.clientX;
                const sliderWidthWithPadding = sliderWidth + sliderContentPadding;
                /* Identify the slider distance from from its wrapper and subtract slider controller width in-order to position the slider controller in center */
                let clickOffset = Math.round(clientX - wrapperRef.current.getBoundingClientRect().left - controllerWidth / 2);
                /*  If we drag the slider out of its  parent boundaries then set clickOffset to its nearest boundary */
                let controllerOffset = clickOffset;

                if (clickOffset > sliderWidthWithPadding) {
                    clickOffset = sliderWidth;
                    controllerOffset = sliderWidthWithPadding;
                } else if (clickOffset < sliderContentPadding) {
                    clickOffset = 0;
                    controllerOffset = sliderContentPadding;
                }
                sliderControllerRef.current.style.left = `${controllerOffset}px`;
                tableRef.current.scroll(clickOffset * oneMinimapToTableWidth, wrapperRef.current.offsetTop);
            },
            [wrapperRef.current, controllerWidth, oneMinimapToTableWidth, tableRef.current, sliderWidth]
        ),
        onSliderControllerMove = useCallback((e: any) => mouseDown && positionSliderController(e), [mouseDown, positionSliderController]),
        onSliderControllerMouseDown = useCallback(
            (e: React.MouseEvent<HTMLDivElement>) => {
                e.stopPropagation();
                setMouseDown(true);
                positionSliderController(e);
            },
            [positionSliderController]
        ),
        onSliderControllerReset = () => {
            setMouseDown(false);
            window.removeEventListener('mousemove', onSliderControllerMove);
        };

    useEffect(() => {
        setMinimapDimensions();
        window.addEventListener('load', setMinimapDimensions);
        window.addEventListener('resize', setMinimapDimensions);
        return () => {
            window.addEventListener('load', setMinimapDimensions);
            window.removeEventListener('resize', setMinimapDimensions);
        };
    }, [setMinimapDimensions]);

    useEffect(() => {
        const tableInstance = tableRef.current;
        tableInstance.addEventListener('scroll', onTableScroll);
        return () => {
            tableInstance.removeEventListener('scroll', onTableScroll);
        };
    }, [onTableScroll]);

    useEffect(() => {
        window.addEventListener('scroll', onWindowScroll);
        return () => {
            window.removeEventListener('scroll', onWindowScroll);
        };
    }, [onWindowScroll]);

    useEffect(() => {
        window.addEventListener('mousemove', onSliderControllerMove);
        return () => {
            window.removeEventListener('mousemove', onSliderControllerMove);
        };
    }, [onSliderControllerMove]);

    useEffect(() => {
        window.addEventListener('mouseup', onSliderControllerReset);
        document.documentElement.addEventListener('mouseleave', onSliderControllerReset);
        return () => {
            window.removeEventListener('mouseup', onSliderControllerReset);
            document.documentElement.removeEventListener('mouseleave', onSliderControllerReset);
        };
    }, [onSliderControllerReset]);

    return (
        isHorizontalScrollPresent && (
            <MinimapWrapper ref={wrapperRef} sliderWidth={minimapWidth} offsetTop={minimapOffsetTop}>
                <StyledMinimap onMouseDown={onSliderControllerMouseDown} sliderWidth={minimapWidth}>
                    {sliderContent}
                    <SliderController
                        onMouseDown={onSliderControllerMouseDown}
                        ref={sliderControllerRef}
                        controllerWidth={controllerWidth}
                        leftOffset={sliderContentPadding}
                    />
                </StyledMinimap>
            </MinimapWrapper>
        )
    );
});

Minimap.defaultProps = {
    minimapWidth: 130,
    controllerWidth: 30,
    sliderContentPadding: 8
};
