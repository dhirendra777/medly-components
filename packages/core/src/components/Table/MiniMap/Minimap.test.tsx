import { render } from '@test-utils';
import React from 'react';
import { Minimap } from './Minimap';

const renderMinimap = (props?: { scrollWidth: number; offsetWidth: number }) => {
    const { scrollWidth = 300, offsetWidth = 200 } = props || {};
    const tableRef = {
        current: {
            scrollWidth,
            offsetWidth,
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            getBoundingClientRect: jest.fn().mockImplementation(() => ({
                bottom: 100
            }))
        }
    };
    return { tableRef, ...render(<Minimap tableRef={tableRef} />) };
};

describe('Minimap component', () => {
    it('should render properly', () => {
        jest.spyOn(document.documentElement, 'getBoundingClientRect').mockImplementation(
            () =>
                ({
                    height: 80,
                    y: 0
                } as DOMRect)
        );
        const { container } = renderMinimap();
        expect(container).toMatchSnapshot();
    });
});
