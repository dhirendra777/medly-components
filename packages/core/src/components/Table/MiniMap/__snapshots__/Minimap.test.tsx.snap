// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minimap component should render properly 1`] = `
.c0 {
  position: -webkit-sticky;
  position: sticky;
  display: block;
  z-index: 999;
  left: NaNpx;
  width: 130px;
  bottom: 115px;
  opacity: 0.4;
  -webkit-transition: opacity 400ms ease-in-out;
  transition: opacity 400ms ease-in-out;
}

.c0:hover {
  opacity: 1;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
}

.c1 {
  position: absolute;
  display: block;
  left: 0;
  background: #ffffff;
  height: 4.6rem;
  box-shadow: 0 .4rem .6rem .3rem rgba(51,51,51,0.1);
  z-index: 999;
  border-radius: .4rem;
}

.c3 {
  width: 30px;
  cursor: pointer;
  margin: .8rem .25rem;
  height: 3rem;
  background: #dfe4e9;
  z-index: 999;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 130px;
  z-index: 999;
}

.c2 > div:nth-child(1) {
  margin-left: .8rem;
}

.c2 > div:last-child {
  margin-right: .8rem;
}

.c6 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c6 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c4 {
  width: 30px;
  left: 8px;
  cursor: pointer;
  height: 4.6rem;
  padding: 1.3rem 0;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  top: 0;
  z-index: 9999;
}

.c4 .c5 * {
  fill: #126AFA;
}

.c4:hover .c5 * {
  fill: #0D4AAF;
}

.c4:active .c5 * {
  fill: #09357D;
}

.c4 .c5 {
  position: absolute;
  left: -.4rem;
}

.c4 .c5:nth-child(2) {
  left: 1rem;
}

<div>
  <tr
    class="c0"
  >
    <td
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        />
        <div
          class="c3"
        />
        <div
          class="c3"
        />
        <div
          class="c3"
        />
        <div
          class="c3"
        />
      </div>
      <div
        class="c4"
      >
        <svg
          class="c5 c6"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.791 11.005H7.621l4.88-4.88c.39-.39.39-1.03 0-1.42a.996.996 0 0 0-1.41 0l-6.59 6.59a.996.996 0 0 0 0 1.41l6.59 6.59a.996.996 0 1 0 1.41-1.41l-4.88-4.88h11.17c.55 0 1-.45 1-1s-.45-1-1-1z"
            fill="#000"
            fill-opacity=".54"
          />
        </svg>
        <svg
          class="c5 c6"
          fill="none"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.209 13h11.17l-4.88 4.88c-.39.39-.39 1.03 0 1.42.39.39 1.02.39 1.41 0l6.59-6.59a.996.996 0 0 0 0-1.41l-6.58-6.6a.996.996 0 0 0-1.41 0 .996.996 0 0 0 0 1.41l4.87 4.89H5.209c-.55 0-1 .45-1 1s.45 1 1 1z"
            fill="#000"
            fill-opacity=".54"
          />
        </svg>
      </div>
    </td>
  </tr>
</div>
`;
